File: Hotelio.PromoCodeService/MonolithRestClients/UserClient.cs
namespace Hotelio.PromoCodeService.MonolithClients
{
    public interface IUserClient
    {
        Task<bool> IsUserActive(string userId);
        Task<bool> IsUserBlacklisted(string userId);
        Task<string?> GetUserStatus(string userId);
    }

    public class UserClient : IUserClient
    {
        private readonly HttpClient _httpClient;
        private readonly ILogger<UserClient> _logger;

        public UserClient(HttpClient httpClient, ILogger<UserClient> logger)
        {
            _httpClient = httpClient;
            _logger = logger;
        }

        public async Task<bool> IsUserActive(string userId)
        {
            try
            {
                var response = await _httpClient.GetAsync($"/api/users/{userId}/active");
                response.EnsureSuccessStatusCode();

                var content = await response.Content.ReadAsStringAsync();
                return bool.Parse(content);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to check if user {UserId} is active", userId);
                throw;
            }
        }

        public async Task<bool> IsUserBlacklisted(string userId)
        {
            try
            {
                var response = await _httpClient.GetAsync($"/api/users/{userId}/blacklisted");
                response.EnsureSuccessStatusCode();

                var content = await response.Content.ReadAsStringAsync();
                return bool.Parse(content);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to check if user {UserId} is blacklisted", userId);
                throw;
            }
        }

        public async Task<string?> GetUserStatus(string userId)
        {
            try
            {
                return await _httpClient.GetStringAsync($"/api/users/{userId}/status");
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Failed to get status for user {UserId}", userId);
                return null;
            }
        }
    }
}

File: microservices/booking-service/rest-clients/userClient.js
const axios = require('axios');

const USER_SERVICE_URL = process.env.USER_SERVICE_URL || 'http://monolith:8080';

class UserClient {
    async getUserStatus(userId) {
        try {
            const response = await axios.get(`${USER_SERVICE_URL}/api/users/${userId}/status`);
            return response.data;
        } catch (error) {
            throw new Error(`Failed to get user status: ${error.message}`);
        }
    }

    async isUserActive(userId) {
        try {
            const response = await axios.get(`${USER_SERVICE_URL}/api/users/${userId}/active`);
            return response.data === true;
        } catch (error) {
            throw new Error(`Failed to check user active: ${error.message}`);
        }
    }

    async isUserBlacklisted(userId) {
        try {
            const response = await axios.get(`${USER_SERVICE_URL}/api/users/${userId}/blacklisted`);
            return response.data === true;
        } catch (error) {
            throw new Error(`Failed to check user blacklisted: ${error.message}`);
        }
    }

    async isUserVip(userId) {
        try {
            const response = await axios.get(`${USER_SERVICE_URL}/api/users/${userId}/vip`);
            return response.data === true;
        } catch (error) {
            return false;
        }
    }

    async getBasePrice(userId) {
        const isVip = await this.isUserVip(userId);
        return isVip ? 80.0 : 100.0;
    }
}

module.exports = new UserClient();